src/dialog-controller.js

```diff
@@ -31,19 +29,16 @@ export class DialogController {
 
   close(ok: boolean, result: any) {
     let returnResult = new DialogResult(!ok, result);
-    return invokeLifecycle(this.viewModel, 'canDeactivate').then(canDeactivate => {
-      if (canDeactivate) {
-        return invokeLifecycle(this.viewModel, 'deactivate')
+    return invokeLifecycle(this.viewModel, 'canDeactivate')
+      .then(canDeactivate => !canDeactivate ? null :
+        invokeLifecycle(this.viewModel, 'deactivate')
+          .then(() => this._renderer.hideDialog(this))
+          .then(() => this._renderer.destroyDialogHost(this))
           .then(() => {
-            return this._renderer.hideDialog(this);
-          }).then(() => {
-            return this._renderer.destroyDialogHost(this);
-          }).then(() => {
             this.controller.unbind();
             this._resolve(returnResult);
-          });
-      }
-    });
+          })
```

Comment:
it's technically not, though I can see how this is confusing. Since I'm using big arrow notation's "compact" mode (I don't know if there's a proper name for this), this whole statement is the "return" of the anon function. The reason I left the paren below this unindented was to show which `then` it was aligned with.